globus_sdk/auth/token_response.py

```diff
@@ -162,32 +161,40 @@ def by_scopes(self):
         """
         return self._by_scopes
 
-    def decode_id_token(self):
+    def decode_id_token(self, openid_configuration=None, jwk=None):
         """
-        A parsed ID Token (OIDC) as a dict.
+        Parse the included ID Token (OIDC) as a dict and return it.
+
+        :param openid_configuration: The OIDC config as a GlobusHTTPResponse or dict.
+            When not provided, it will be fetched automatically.
+        :type openid_configuration: dict or GlobusHTTPResponse
+        :param jwk: The JWK as a cryptography public key object. When not provided, it
+            will be fetched and parsed automatically.
+        :type jwk: RSAPublicKey
         """
-        logger.info('Decoding ID Token "{}"'.format(self["id_token"]))
+        logger.info('Decoding ID Token "%s"', self["id_token"])
         auth_client = self._client
 
-        logger.debug("Fetch JWK Data: Start")
-        oidc_conf = auth_client.get("/.well-known/openid-configuration")
-        jwks_uri = oidc_conf["jwks_uri"]
-        signing_algos = oidc_conf["id_token_signing_alg_values_supported"]
-
-        # use the auth_client's decision on ssl_verify=yes/no
-        jwk_data = requests.get(jwks_uri, verify=auth_client._verify).json()
-        logger.debug("Fetch JWK Data: Complete")
-        # decode from JWK to an RSA PEM key for JWT decoding
-        jwk_as_pem = jwt.algorithms.RSAAlgorithm.from_jwk(
-            json.dumps(jwk_data["keys"][0])
-        )
+        if not openid_configuration:
+            logger.debug("No OIDC Config provided, autofetching...")
+            openid_configuration = auth_client.get_openid_configuration()
+
+        if not jwk:
+            logger.debug("No JWK provided, autofetching + decoding...")
+            jwk = auth_client.get_jwk(
+                openid_configuration=openid_configuration, as_pem=True
+            )
```

Comment:
`.decode_id_token(jwk=jwk_data)` is valid but also sort of strange since it'll pointless fetch the openid config but not use it.

Considering our openid provider uses a jwks that doesn't change I wonder if it doesn't make sense to make `decode_id_token` take the jwks as an optional argument and then either fetch both the config/jwks or nothing? Do you think there's a scenario in which an SDK consumer would try to optimize by fetching only the openid config but not the jwks?

Reply:
I did consider making a named tuple or something like `JWTRequirements.openid_configuration` and `JWTRequirements.jwk`, and only allowing
- `AuthClient.get_openid_configuration`
- `AuthClient.get_jwt_requirements`
- `OAuthTokenResponse.decode_id_token(jwt_requirements=...)`

Or we could make it an error to pass the JWK without passing openid configuration (but not vice versa).

If you want to do anything about this, I think it's one of those two? I'm a bigger fan of the usage error than a new type that wraps the pair of data.

Reply:
I went ahead and added a commit which makes it an error to pass `jwk` without `openid_configuration`.

If this looks like a reasonable direction, I'd like to move ahead and build out tests for it and un-mark it as a draft when I have time. Let me know if the general bent of this seems okay. (We can work out details later.)